File: service/db/mysql/training.py
import io
import json
import time

import mysql.connector


from service.logging import BackendLogger

db_logger = BackendLogger("db")


class TrainingRepository:
    def __init__(self, connection):
        self.connection = connection

    def get_connection(self):
        return self.connection.get_connection()

    def _mutate_training_games(self, batch_id: str, mutate,
                               *, current_game_id=None,
                               current_generation=None) -> bool:
        """Serialize read-modify-write updates to the batch games JSON array."""
        conn = self.get_connection()
        if not conn:
            return False
        cursor = conn.cursor(dictionary=True)
        try:
            cursor.execute(
                "SELECT games FROM training_batches "
                "WHERE batch_id = %s FOR UPDATE",
                (batch_id,),
            )
            row = cursor.fetchone()
            if not row:
                conn.rollback()
                return False
            games = row.get("games", []) or []
            if isinstance(games, str):
                games = json.loads(games)
            mutate(games)

            clauses = ["games = %s"]
            params: list[object] = [json.dumps(games)]
            if current_game_id is not None:
                clauses.extend([
                    "status = 'running'",
                    "current_game_id = %s",
                    "current_generation = %s",
                ])
                params.extend([current_game_id, current_generation])
            params.append(batch_id)
            cursor.execute(
                f"UPDATE training_batches SET {', '.join(clauses)} "
                "WHERE batch_id = %s",
                tuple(params),
            )
            conn.commit()
            return True
        except (mysql.connector.Error, json.JSONDecodeError) as err:
            conn.rollback()
            db_logger.error(f"Error updating training batch games: {err}")
            return False
        finally:
            cursor.close()
            conn.close()

    def create_training_batch(self, batch_id: str, config: dict) -> bool:
        conn = self.get_connection()
        if not conn:
            raise ConnectionError("Could not connect to MySQL")
        cursor = conn.cursor()
        query = """
            INSERT INTO training_batches
            (batch_id, status, ruleset, bot_model, bot_count, total_generations,
             current_generation, last_heartbeat_at, phase, base_genome_id,
             config, generation_statistics, games)
            VALUES (%s, 'running', %s, %s, %s, %s, 0, NOW(), 'pending', %s, %s, %s, %s)
        """
        try:
            cursor.execute(query, (
                batch_id, config.get("ruleset"), config.get("bot_model"),
                config.get("bot_count"), config.get("total_generations"),
                config.get("base_genome_id"), json.dumps(config.get("config", {})),
                json.dumps([]), json.dumps([])))
            conn.commit()
            return True
        except mysql.connector.Error as err:
            conn.rollback()
            db_logger.error(f"Error creating training batch: {err}")
            raise
        finally:
            cursor.close()
            conn.close()

    def mark_training_batch_game_started(self, batch_id: str, game_id: str,
                                         generation: int,
                                         attempt: int | None = None):
        def append_game(games):
            games.append({
                "game_id": game_id,
                "generation": generation,
                "attempt": attempt,
                "status": "spawning",
                "error_message": None,
                "genome_count": 0,
                "best_fitness": None,
                "average_fitness": None,
            })

        return self._mutate_training_games(
            batch_id, append_game,
            current_game_id=game_id,
            current_generation=generation,
        )

    def mark_training_batch_game_running(self, batch_id: str, game_id: str):
        def mark_running(games):
            for game in games:
                if game.get("game_id") == game_id:
                    game["status"] = "running"
                    break

        return self._mutate_training_games(batch_id, mark_running)

    def mark_training_batch_game_failed(self, batch_id: str, game_id: str,
                                        error_message: str):
        def mark_failed(games):
            for game in games:
                if game.get("game_id") == game_id:
                    game["status"] = "failed"
                    game["error_message"] = error_message
                    game["genome_count"] = int(
                        game.get("genome_count", 0) or 0)
                    break

        return self._mutate_training_games(batch_id, mark_failed)

    def mark_training_batch_game_completed(self, batch_id: str, game_id: str,
                                           genome_count: int,
                                           fitness_summary: dict | None = None):
        summary = fitness_summary or {}

        def mark_completed(games):
            for game in games:
                if game.get("game_id") == game_id:
                    game["status"] = "completed"
                    game["error_message"] = None
                    game["genome_count"] = int(genome_count or 0)
                    game["best_fitness"] = summary.get("best_fitness")
                    game["average_fitness"] = summary.get("average_fitness")
                    break

        return self._mutate_training_games(batch_id, mark_completed)

    def record_training_batch_heartbeat(self, batch_id: str, phase: str,
                                        current_generation: int,
                                        current_game_id: str | None = None):
        conn = self.get_connection()
        if not conn:
            return
        cursor = conn.cursor()
        try:
            cursor.execute(
                """
                UPDATE training_batches
                SET last_heartbeat_at = NOW(), phase = %s,
                    current_generation = %s, current_game_id = %s
                WHERE batch_id = %s
                """,
                (phase, current_generation, current_game_id, batch_id),
            )
            conn.commit()
        except mysql.connector.Error as err:
            db_logger.error(f"Error recording training batch heartbeat: {err}")
        finally:
            cursor.close()
            conn.close()

    def update_training_batch_status(self, batch_id: str, status: str,
                                     error_message: str | None = None):
        conn = self.get_connection()
        if not conn:
            return
        cursor = conn.cursor()
        completed_clause = (
            ", completed_at = NOW()"
            if status in ("completed", "failed", "cancelled", "stalled")
            else ""
        )
        try:
            cursor.execute(
                f"""
                UPDATE training_batches
                SET status = %s, last_error = %s{completed_clause}
                WHERE batch_id = %s
                """,
                (status, error_message, batch_id),
            )
            conn.commit()
        except mysql.connector.Error as err:
            db_logger.error(f"Error updating training batch status: {err}")
        finally:
            cursor.close()
            conn.close()

    def append_training_batch_generation_stats(self, batch_id: str, stats: dict):
        batch = self.get_training_batch(batch_id)
        if not batch:
            return
        generation_statistics = batch.get("generation_statistics", [])
        generation_statistics.append(stats)
        conn = self.get_connection()
        if not conn:
            return
        cursor = conn.cursor()
        try:
            cursor.execute(
                "UPDATE training_batches SET generation_statistics = %s WHERE batch_id = %s",
                (json.dumps(generation_statistics), batch_id))
            conn.commit()
        except mysql.connector.Error as err:
            db_logger.error(f"Error appending training batch stats: {err}")
        finally:
            cursor.close()
            conn.close()

    def complete_training_batch(self, batch_id: str,
                                final_champion_genome_id: str | None = None):
        conn = self.get_connection()
        if not conn:
            return
        cursor = conn.cursor()
        query = """
            UPDATE training_batches
            SET status = 'completed', completed_at = NOW(),
                final_champion_genome_id = %s
            WHERE batch_id = %s
        """
        try:
            cursor.execute(query, (final_champion_genome_id, batch_id))
            conn.commit()
        except mysql.connector.Error as err:
            db_logger.error(f"Error completing training batch: {err}")
        finally:
            cursor.close()
            conn.close()

    def get_training_batches(self) -> list:
        conn = self.get_connection()
        if not conn:
            return []
        cursor = conn.cursor(dictionary=True)
        try:
            cursor.execute("SELECT * FROM training_batches ORDER BY started_at DESC")
            return [self._decode_training_batch(row) for row in cursor.fetchall()]
        except mysql.connector.Error as err:
            db_logger.error(f"Error getting training batches: {err}")
            return []
        finally:
            cursor.close()
            conn.close()

    def get_training_batch(self, batch_id: str) -> dict | None:
        conn = self.get_connection()
        if not conn:
            return None
        cursor = conn.cursor(dictionary=True)
        try:
            cursor.execute(
                "SELECT * FROM training_batches WHERE batch_id = %s LIMIT 1",
                (batch_id,))
            row = cursor.fetchone()
            return self._decode_training_batch(row) if row else None
        except mysql.connector.Error as err:
            db_logger.error(f"Error getting training batch: {err}")
            return None
        finally:
            cursor.close()
            conn.close()

    def get_training_games(self, batch_id: str) -> list:
        batch = self.get_training_batch(batch_id)
        if not batch:
            return []
        games = batch.get("games", [])
        return games if isinstance(games, list) else []

    def _decode_training_batch(self, row: dict) -> dict:
        for key in ("config", "generation_statistics", "games"):
            if isinstance(row.get(key), str):
                row[key] = json.loads(row[key])
        config = row.get("config", {}) or {}
        games = row.get("games", []) or []
        row["games_per_generation"] = config.get("games_per_generation")
        row["games_completed"] = len([
            game for game in games
            if game.get("status") in ("completed", "failed", "skipped")
        ])
        row["games_failed"] = len([
            game for game in games
            if game.get("status") == "failed"
        ])
        return row

File: service/src/db/mysql/training.ts
import type {
  JsonObject,
  TrainingBatchConfig,
  TrainingBatchRecord,
  TrainingGameRecord,
} from "../contracts.js";
import { decodeJson, rows, type SqlExecutor, type TransactionalSqlExecutor } from "./sql.js";

function decodeBatch(row: Record<string, unknown>): TrainingBatchRecord {
  const config = (decodeJson(row.config) ?? {}) as JsonObject;
  const generationStatistics = (decodeJson(row.generation_statistics) ?? []) as JsonObject[];
  const games = (decodeJson(row.games) ?? []) as TrainingGameRecord[];
  return {
    ...row,
    config,
    generation_statistics: generationStatistics,
    games,
    games_per_generation: typeof config.games_per_generation === "number" ? config.games_per_generation : undefined,
    games_completed: games.filter((game) => ["completed", "failed", "skipped"].includes(game.status)).length,
    games_failed: games.filter((game) => game.status === "failed").length,
  } as TrainingBatchRecord;
}

export class TrainingRepository {
  constructor(private readonly database: TransactionalSqlExecutor) {}

  async createTrainingBatch(batchId: string, config: TrainingBatchConfig): Promise<boolean> {
    await this.database.execute(
      `INSERT INTO training_batches
       (batch_id, status, ruleset, bot_model, bot_count, total_generations, current_generation, last_heartbeat_at, phase, base_genome_id, config, generation_statistics, games)
       VALUES (?, 'running', ?, ?, ?, ?, 0, NOW(), 'pending', ?, ?, ?, ?)`,
      [batchId, config.ruleset ?? null, config.bot_model ?? null, config.bot_count ?? null,
        config.total_generations ?? null, config.base_genome_id ?? null, JSON.stringify(config.config ?? {}), "[]", "[]"],
    );
    return true;
  }

  private async mutateGames(
    batchId: string,
    mutate: (games: TrainingGameRecord[]) => void,
    current?: { gameId: string; generation: number },
  ): Promise<void> {
    const connection = await this.database.getConnection?.();
    const executor = connection ?? this.database;
    try {
      await connection?.beginTransaction();
      const batch = await this.loadTrainingBatch(executor, batchId, true);
      if (!batch) {
        await connection?.rollback();
        return;
      }
      mutate(batch.games);
      if (current) {
        await executor.execute(
          "UPDATE training_batches SET games = ?, status = 'running', current_game_id = ?, current_generation = ? WHERE batch_id = ?",
          [JSON.stringify(batch.games), current.gameId, current.generation, batchId],
        );
      } else {
        await executor.execute("UPDATE training_batches SET games = ? WHERE batch_id = ?", [JSON.stringify(batch.games), batchId]);
      }
      await connection?.commit();
    } catch (error) {
      await connection?.rollback();
      throw error;
    } finally {
      connection?.release();
    }
  }

  async markTrainingBatchGameStarted(batchId: string, gameId: string, generation: number, attempt: number | null = null): Promise<void> {
    await this.mutateGames(batchId, (games) => games.push({
      game_id: gameId,
      generation,
      attempt,
      status: "spawning",
      error_message: null,
      genome_count: 0,
      best_fitness: null,
      average_fitness: null,
    }), { gameId, generation });
  }

  async markTrainingBatchGameRunning(batchId: string, gameId: string): Promise<void> {
    await this.mutateGames(batchId, (games) => {
      const game = games.find((candidate) => candidate.game_id === gameId);
      if (game) game.status = "running";
    });
  }

  async markTrainingBatchGameFailed(batchId: string, gameId: string, errorMessage: string): Promise<void> {
    await this.mutateGames(batchId, (games) => {
      const game = games.find((candidate) => candidate.game_id === gameId);
      if (game) {
        game.status = "failed";
        game.error_message = errorMessage;
      }
    });
  }

  async markTrainingBatchGameCompleted(batchId: string, gameId: string, genomeCount: number, fitnessSummary: JsonObject = {}): Promise<void> {
    await this.mutateGames(batchId, (games) => {
      const game = games.find((candidate) => candidate.game_id === gameId);
      if (game) {
        game.status = "completed";
        game.error_message = null;
        game.genome_count = genomeCount;
        game.best_fitness = typeof fitnessSummary.best_fitness === "number" ? fitnessSummary.best_fitness : null;
        game.average_fitness = typeof fitnessSummary.average_fitness === "number" ? fitnessSummary.average_fitness : null;
      }
    });
  }

  async recordTrainingBatchHeartbeat(batchId: string, phase: string, currentGeneration: number, currentGameId: string | null = null): Promise<void> {
    await this.database.execute(
      "UPDATE training_batches SET last_heartbeat_at = NOW(), phase = ?, current_generation = ?, current_game_id = ? WHERE batch_id = ?",
      [phase, currentGeneration, currentGameId, batchId],
    );
  }

  async updateTrainingBatchStatus(batchId: string, status: string, errorMessage: string | null = null): Promise<void> {
    const terminal = ["completed", "failed", "cancelled", "stalled"].includes(status);
    await this.database.execute(
      `UPDATE training_batches SET status = ?, last_error = ?${terminal ? ", completed_at = NOW()" : ""} WHERE batch_id = ?`,
      [status, errorMessage, batchId],
    );
  }

  async appendTrainingBatchGenerationStats(batchId: string, stats: JsonObject): Promise<void> {
    const batch = await this.getTrainingBatch(batchId);
    if (!batch) return;
    batch.generation_statistics.push(stats);
    await this.database.execute(
      "UPDATE training_batches SET generation_statistics = ? WHERE batch_id = ?",
      [JSON.stringify(batch.generation_statistics), batchId],
    );
  }

  async completeTrainingBatch(batchId: string, finalChampionGenomeId: string | null = null): Promise<void> {
    await this.database.execute(
      "UPDATE training_batches SET status = 'completed', completed_at = NOW(), final_champion_genome_id = ? WHERE batch_id = ?",
      [finalChampionGenomeId, batchId],
    );
  }

  async getTrainingBatches(): Promise<TrainingBatchRecord[]> {
    const [result] = await this.database.execute("SELECT * FROM training_batches ORDER BY started_at DESC");
    return rows(result).map(decodeBatch);
  }

  async getTrainingBatch(batchId: string, forUpdate = false): Promise<TrainingBatchRecord | null> {
    return this.loadTrainingBatch(this.database, batchId, forUpdate);
  }

  private async loadTrainingBatch(executor: SqlExecutor, batchId: string, forUpdate = false): Promise<TrainingBatchRecord | null> {
    const [result] = await executor.execute(
      `SELECT * FROM training_batches WHERE batch_id = ? LIMIT 1${forUpdate ? " FOR UPDATE" : ""}`,
      [batchId],
    );
    const row = rows(result)[0];
    return row ? decodeBatch(row) : null;
  }

  async getTrainingGames(batchId: string): Promise<TrainingGameRecord[]> {
    return (await this.getTrainingBatch(batchId))?.games ?? [];
  }
}
